### PR TITLE
feat: [#2856] Emit actionstart and actioncomplete events on actors

### LINKED ISSUE #2856: Emit "actionstart" and "actioncomplete" events
It would be useful the ActionsComponent emitted events when starting/completing actions

### Context

I have a moving platform that has its path defined by actions

```ts
    const platform = new MovingPlatform(
      {
        x: 300,
        y: 50,
        width: 100,
        height: 16,
      },
      (actions) =>
        actions.repeatForever((builder) =>
          builder.moveBy(0, -50, 40).moveBy(0, 50, 40)
        )
    )
```

I have some particular logic that I need to run whenever the platform changes direction that should happen immediately after the velocity is changed. Hooking into an event would be a very nice way to solve this.

### Proposal

Action component the following events containing the relevant action:

- `actionstart`
- `actioncomplete`

These can be either on the actor or the actions component, or both? This would also fire for each action in `repeatForever` etc.

```ts
type ActionStartEvent {
    action: ex.Action
}

type ActionCompleteEvent {
   action: ex.Action
}

this.on('actioncomplete', (ev) => {
   
})
```

### PR DESCRIPTION
**Pull request by mattjennings**
===:clipboard: PR Checklist :clipboard:===

- [x] :pushpin: issue exists in github for these changes
- [x] :microscope: existing tests still pass
- [x] :see_no_evil: code conforms to the [style guide](https://github.com/excaliburjs/Excalibur/blob/main/STYLEGUIDE.md)
- [x] :triangular_ruler: new tests written and passing / old tests updated with new scenario(s)
- [x] :page_facing_up: changelog entry added (or not needed)

==================

Closes #2856

## Changes:

- Added `actionstart` and `actioncomplete` events to the Actor that are fired when an action starts and completes

### CODE REVIEW

File: src/engine/Events.ts [hunk 1]
```diff
@@ -545,3 +549,21 @@ export class ExitTriggerEvent extends GameEvent<Actor> {
     super();
   }
 }
+
+/**
+ * Event thrown on an [[Actor]] when an action starts.
+ */
+export class ActionStartEvent extends GameEvent<Entity> {
```
Comment (mattjennings) [inline src/engine/Events.ts:556]: @eonarheim I wasn't too sure if this should be `GameEvent<Entity>` or if you'd have wanted an interface to be made so I could do the `<T extends XYZ = Entity>` trick?

Comment (eonarheim) [reply src/engine/Events.ts:556]: @mattjennings This is perfect for now! Fits with the current pattern. 

Eventually I want to rethink GameEvent entirely (maybe remove it?) but I haven't put enough effort into what the future would look like.

**Review by eonarheim — APPROVED** [review]

**Comment by eonarheim** [comment]
Looks like a test failure, not sure if this is flakey yet

File: src/engine/Events.ts (hunk 1 above)
Comment (mattjennings) [reply src/engine/Events.ts:556]: sounds good! 

It's pretty nice having these to use as types when setting up handlers on a class:

```ts
onInitialize() {
     this.on('actionstart', this.onActionStart)
}

onActionStart(ev: ex.ActionStartEvent) { ... }
```

so I'd advocate for keeping them

**Comment by mattjennings** [comment]
@eonarheim I think it was legitimate - just pushed a change and it's awaiting approval to run again

**Review by eonarheim — APPROVED** [review]

**Comment by mattjennings** [comment]
@eonarheim I wonder if there should be an `actionupdate` event too?

**Comment by eonarheim** [comment]
> @eonarheim I wonder if there should be an `actionupdate` event too?

I suppose it might be useful to know what action is running, although start and complete will give you that. Perhaps useful to peer into the current actions state?

**Comment by eonarheim** [comment]
I'm okay without an update for now

**Comment by mattjennings** [comment]
```ts
    this.actions.moveBy(0, 50, 40) 

    this.on('actionstart', (ev) => {
      console.log(this.vel)
    })

    this.on('actioncomplete', (ev) => {
      console.log(this.vel)
    })
```

In this example, `this.vel` will be 0,0 both times as it either hasn't run its first update yet or is completed already. For this use-case though I could just store the vel from the actor's update loop and then compare in the actioncomplete.

I think I'm fine without it too until convinced of a necessary use-case.

**Comment by eonarheim** [comment]
@mattjennings Thanks again!

**Pull request merged**